### PR TITLE
Cache: init empty cache when `use_cache`

### LINKED ISSUE #34206: MllamaForCausalLM not returning past_key_values even with use_cache=True
### System Info

- `transformers` version: 4.45.2
- Platform: Linux-5.4.0-187-generic-x86_64-with-glibc2.31
- Python version: 3.11.5
- Huggingface_hub version: 0.23.4
- Safetensors version: 0.4.3
- Accelerate version: 0.33.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.1.2 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA RTX 6000 Ada Generation

### Who can help?

@amyeroberts @ArthurZucker 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

```python
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch
import torch.nn as nn
import os
import matplotlib.pyplot as plt
from promptmix import generate_probmix
from transformers import MllamaForConditionalGeneration, AutoProcessor


model_id = "meta-llama/Llama-3.2-11B-Vision-Instruct" # https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct


tokenizer = AutoTokenizer.from_pretrained(model_id)

model = AutoModelForCausalLM.from_pretrained(
    model_id, 
    cache_dir=cache_directory,
    torch_dtype=torch.float16,
    device_map='auto',
    )

processor = AutoProcessor.from_pretrained(model_id, cache_dir=cache_directory)

input_ids = tokenizer.encode('Hi, tell me a story of frog.', add_special_tokens=False, return_tensors='pt').to(model.device)

with torch.no_grad():
    output = model.forward(input_ids=input_ids, use_cache=True)
output
```

### Expected behavior

I expect to see a `past_key_values` in the `output`. However, I got `None`.

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/34206. As per title we would have to initialize empty cache whenever `use_cache=True`. Additionally `MllamaForCausalLM` was not loading correctly for me, so I modified the base model prefix

### CODE REVIEW

File: tests/generation/test_utils.py [hunk 1]
```diff
@@ -1524,6 +1524,14 @@ def test_past_key_values_format(self):
...
             per_head_embed_dim = embed_dim // num_attention_heads
 
+            # some models have diffent num-head for query vs key/value so we need to assign correct value
+            # BUT only after `per_head_embed_dim` is set
+            num_attention_heads = (
+                text_config.num_key_value_heads
+                if getattr(text_config, "num_key_value_heads", None) is not None
+                else num_attention_heads
+            )
+
```
Comment (gante) [inline tests/generation/test_utils.py:1534]: Make sure `py.test tests/models -k test_past_key_values_format` passes :)

File: tests/models/qwen2_vl/test_modeling_qwen2_vl.py [hunk 2]
```diff
@@ -245,6 +246,91 @@ def test_initialization(self):
                         msg=f"Parameter {name} of model {model_class} seems not properly initialized",
                     )
 
+    @pytest.mark.generate
+    def test_generate_continue_from_past_key_values(self):
```
Comment (gante) [inline tests/models/qwen2_vl/test_modeling_qwen2_vl.py:250]: If the missing `rope_deltas` is the only issue, can we instead:
1 - replace `prepare_config_and_inputs_for_common()` by `prepare_config_and_inputs_for_generate()` in the base test
2 - Overload `prepare_config_and_inputs_for_generate` in this model tester, to also return `rope_deltas`

(much smaller code diff -> easier to maintain)

**Review by gante — COMMENTED** [review]
Two items to check :D

File: src/transformers/models/mllama/modeling_mllama.py [hunk 3]
```diff
@@ -1838,7 +1841,7 @@ def _prepare_4d_causal_attention_mask_with_cache_position(
 )
 class MllamaForCausalLM(MllamaPreTrainedModel, GenerationMixin):
     config_class = MllamaTextConfig
-    base_model_prefix = "model"
+    base_model_prefix = "language_model"
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1844]: Yes, that is how the checkpoints looks like and when I loaded the model it didn't load correctly if the base prefix is not fixed. The slow tests unfortunately can't be run because the model is read-token protected and EU has no access to them 🥲  But I tested with open mirrored weights

File: tests/models/qwen2_vl/test_modeling_qwen2_vl.py (hunk 2 above)
Comment (zucchini-nlp) [reply tests/models/qwen2_vl/test_modeling_qwen2_vl.py:250]: Hmm, but we would need access to the model outputs from prev step in that case, which we don't have because `generate()` returns only some kwargs. And if we want to feed the original inputs with cache to the `prepare_config_and_inputs_for_generate`, we'll need to prepare `repo_deltas` again in the same way

File: tests/generation/test_utils.py (hunk 1 above)
Comment (zucchini-nlp) [reply tests/generation/test_utils.py:1534]: passed

File: src/transformers/models/mllama/modeling_mllama.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/models/mllama/modeling_mllama.py:1844]: Fair :)

Can we change `self.model` to `self.language_model`? Some parts of the codebase call variants of `inner_model = getattr(model, model.base_model_prefix)`

[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/models/mllama/modeling_mllama.py:1844]: > that would mean we change the checkpoint state dict keys right?

Uhmmm possibly? Not sure 😅 The tests that save and reload would break if it would be BC, I think

File: tests/models/qwen2_vl/test_modeling_qwen2_vl.py (hunk 2 above)
Comment (gante) [reply tests/models/qwen2_vl/test_modeling_qwen2_vl.py:250]: (chatted on slack about exploring how to standardize `qwen2_vl` + `generate`)

File: src/transformers/models/mllama/modeling_mllama.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1844]: Indeed, I found why it was changed, it was to get the self.base_model method for preparing causal mask in https://github.com/huggingface/transformers/pull/33677 hehe

and yes, w/o changing state dict keys we cannot call it "model". Imo even if we change the official state dict, there are many mirrors/finetunes which will be BC breaking compared to the first model release. So the better way i think is to bring back the base-model-prefix as it was.

I'm thinking maybe we can have a default method for `update_causal_mask` and `preapre_attention_mask` which will be the fallback if the base-model has no such method defined? 🤔

[the remark this answers is not in the document]
Comment (gante) [reply src/transformers/models/mllama/modeling_mllama.py:1844]: 🤔 

Regarding the `generate`-specific problem: if `base_model = getattr(self, self.base_model_prefix, None)` in the generalist `prepare_inputs_for_generation` (or its downstream usage) is the issue, then my recommendation would be to overwrite `prepare_inputs_for_generation` in mllama -- more specifically, in the classes where it doesn't work 

Alternatively, we could define `_prepare_4d_causal_attention_mask_with_cache_position` in all model classes -- write it once in the innermost class, then the child classes would define this function as a parent call.

Any of these solutions would work well for me :) (with a preference for the second: when we rewrite `prepare_inputs_for_generation`, we know for sure we'll have extra maintenance in the future)

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1845]: done, and also enabled compile tests for the CausalLM class to test that it works

File: src/transformers/models/mllama/modeling_mllama.py [hunk 4]
```diff
@@ -1618,6 +1741,9 @@ def forward(
 
         hidden_states = inputs_embeds
 
+        if use_cache and past_key_values is None:
+            past_key_values = DynamicCache()
+
```
Comment (ArthurZucker) [inline src/transformers/models/mllama/modeling_mllama.py:1746]: this does make sense as it's helping users, and an old APi, but let's promote init of a cache and passing it! 🤗

**Review by ArthurZucker — COMMENTED** [review]
Thanks LGTM but let's separate unrelated changes!

File: src/transformers/models/mllama/modeling_mllama.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1746]: We could but still I think it is a lot easier if users want a forward pass with cache, and do not want extra lines of code for importing and passing the cache object. So i think we'd better keep the default cache for now

File: src/transformers/models/mllama/modeling_mllama.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/mllama/modeling_mllama.py:1845]: oke, will make a new PR

**Review by ArthurZucker — APPROVED** [review]
Thanks 🤗
